src/streams_reader.c

```diff
@@ -5,39 +5,270 @@
+
+#define StreamIdIsZero(id) (id.first == 0 && id.second == 0)
+
+static int StreamReader_StreamIdCompare(StreamId id1, StreamId id2){
+    if(id1.second > id2.second){
+        return 1;
+    }
+
+    if(id1.second < id2.second){
+        return -1;
+    }
+
+    if(id1.first > id2.first){
+        return 1;
+    }
+
+    if(id1.first < id2.first){
+        return -1;
+    }
+
+    return 0;
+}
+
+static StreamId StreamReader_ParseStreamId(const char* streamId){
+    StreamId ret;
+    int res = sscanf(streamId, "%ld-%ld", &ret.second, &ret.first);
+    assert(res == 2);
+    return ret;
+}
+
+StreamReaderTriggerArgs* StreamReaderTriggerArgs_Create(const char* streamName, size_t batchSize, size_t durationMS){
+    StreamReaderTriggerArgs* readerArgs = RG_ALLOC(sizeof(StreamReaderTriggerArgs));
+    readerArgs->stream = RG_STRDUP(streamName);
+    readerArgs->batchSize = batchSize;
+    readerArgs->durationMS = durationMS;
+    return readerArgs;
+}
+
+static void StreamReaderTriggerArgs_Free(StreamReaderTriggerArgs* args){
+    RG_FREE(args->stream);
+    RG_FREE(args);
+}
+
+static StreamReaderTriggerCtx* StreamReaderTriggerCtx_GetShallowCopy(StreamReaderTriggerCtx* srtctx){
+    ++srtctx->refCount;
+    return srtctx;
+}
+
+static void StreamReader_ReadLastId(RedisModuleCtx *rctx, SingleStreamReaderCtx* ssrctx){
+    RedisModuleCallReply *reply = RedisModule_Call(rctx, "XINFO", "cc", "STREAM", ssrctx->keyName);
+    bool ret = StreamReader_VerifyCallReply(rctx, reply, "Failed on XINFO command", "warning");
+    assert(ret);
+    assert(RedisModule_CallReplyType(reply) == REDISMODULE_REPLY_ARRAY);
+    assert(RedisModule_CallReplyLength(reply) > 10);
+    RedisModuleCallReply *idReply = RedisModule_CallReplyArrayElement(reply, 9);
+    const char* idStr = RedisModule_CallReplyStringPtr(idReply, NULL);
+    ssrctx->lastId = StreamReader_ParseStreamId(idStr);
+}
+
+#define GEARS_CONSUMER_GROUP "__gears_consumer_group__"
+static void StreamReader_CreateConsumersGroup(RedisModuleCtx *ctx, const char* key){
+    RedisModuleCallReply *r = RedisModule_Call(ctx, "XGROUP", "!cccc", "CREATE", key, GEARS_CONSUMER_GROUP, "0");
+    // if creation fails its ok, this is why we do not call VerifyCallReply
+}
+
+static SingleStreamReaderCtx* SingleStreamReaderCtx_Create(RedisModuleCtx* ctx,
+                                                             const char* keyName,
+                                                             StreamReaderTriggerCtx* srtctx){
+    SingleStreamReaderCtx* ssrctx = RG_ALLOC(sizeof(*ssrctx));
+    ssrctx->numTriggered = 0;
+    ssrctx->srtctx = srtctx;
+    ssrctx->keyName = RG_STRDUP(keyName);
+    ssrctx->timerIsSet = false;
+    StreamReader_ReadLastId(ctx, ssrctx);
+    Gears_dictAdd(srtctx->singleStreamData, (char*)keyName, ssrctx);
+    return ssrctx;
+}
+
+static void StreamReaderTriggerCtx_Free(StreamReaderTriggerCtx* srtctx){
+    if((--srtctx->refCount) == 0){
```

Comment:
Check for !srtctx->refCount?

Reply:
I think its better this way cause its actual number and not a true/false value..

Reply:
Assert(srtctx->refCount != 0)